### ISSUE TITLE
Support for zendframework/zf1-extras

### ISSUE THREAD

**Issue by fredericgboutin-yapla** [zf1s/zf1#128]
Hi,
I'm digging my way around and I'm struggling to find any info in regards to support for "zf1-extras" or "ZendX" components like JQuery - https://framework.zend.com/manual/1.12/en/zendx.jquery.html

I found this fork - https://github.com/Shardj/zf1-extras-future - which seems "popular" so I suppose I could use it in conjunction with zf1s but I would like to hear your input on the whole matter.

Thanks!

**Comment by fredericgboutin-yapla** [comment]
Well, it's been 2 weeks and I've got no feedback. 

Anyway, I went ahead and I tried to use zf1s with zf1-extras and there are 2 problems,

1. There are a lot of `require_once` in zf1-extras codebase which don't resolve with zf1s
2. zf1s is NOT a drop-in replacement for zf1, as it does NOT / cannot inherently specify in the `composer.json` a definition like,

```json
    "replace": {
        "zendframework/zendframework1": ">=1.12.20"
    }
```

All this mean that we would need to fork zf1-extras, remove the `require_once` statements, adjust the code to make it work including the unit tests, put those tests under CI for php compatibility review, etc.

Basically when someone uses zf1s it alienates itself from using other components which could provide features to zf1, for example,

- https://github.com/PHP-DI/ZF1-Bridge
- https://github.com/guilhermeblanco/zendframework1-doctrine2
- https://github.com/jokkedk/ZFDebug
- https://github.com/Shardj/zf1-extras-future

**Comment by falkenhawk** [comment]
@fredericgboutin-yapla now full `zf1s/zf1` package is available. https://packagist.org/packages/zf1s/zf1 v1.15.0 - please try.

**Comment by falkenhawk** [comment]
Thank you for sharing your findings @fredericgboutin-yapla !

Those two classes `Zend_Tool_Framework_Provider_DocblockManifestInterface` and `Zend_Gdata_Analytics_Goal` deserve a fix to rename/move their files so that composer autoloader does not complain.

As for php extensions listed in `require` section.. you might be right, it is probably too restrictive at the moment. Due to the nature of zf1, you may never need some of the extensions, as long as you do not use components which require them. But when you start using them, those components themselves should throw errors that required extensions are not installed.

@glensc did great job in https://github.com/zf1s/zf1/pull/6, filling the requirements for each component individually. Then right after that PR got merged, I went ahead and added them to the main `composer.json` in https://github.com/zf1s/zf1/commit/fdf41f38e332f5879341d7312715785962b27ef3, without a PR, but pushing directly to `master` 🙈 Why, I don't remember anymore.

I suppose they could be added to `suggest` section instead, or dropped entirely from the main `composer.json` 🤔

**Comment by glensc** [comment]
if you remove extension require from root composer.json then our CI can't validate what extensions are required. otoh, the requirements will likely never change and they already are present in CI configs. but if the base image changes extensions included, our jobs have no way to validate what are required. and some tests are just silently skipped if extension is missing and we might not notice we are skipping some tests.

so, mixed feelings here again :)

but of course we could do some hacking like transform all suggests to require in our ci job. something similar done here:
- https://github.com/eventum/eventum/pull/1291

**Comment by glensc** [comment]
ok. moving to require-dev is better, I didn't think of that:
- https://github.com/zf1s/zf1/pull/136

**Comment by fredericgboutin-yapla** [comment]
> As for php extensions listed in require section.. you might be right, it is probably too restrictive at the moment. Due to the nature of zf1, you may never need some of the extensions,

To be fair, I personally think it is alright. The workaround I came with is good because I had to explictly acknowledge the requirement and deal with it somehow. I basically like being told of extension requirements.

But yeah, if I had to deal with half a dozen... and figuring transitive dependencies between different Zend modules to understand what I'm really entitled to. But still, I think it is probably a fair drawback of using the monorepo at the first place.

**Comment by fredericgboutin-yapla** [comment]
@falkenhawk I had to bootstrap zfdebug from https://github.com/jokkedk/ZFDebug with zf1s and I think I basically found something interesting. I truly had to understand what `require_once` is,

>>>
Files are included based on the file path given or, if none is given, the [include_path](https://www.php.net/manual/en/ini.core.php#ini.include-path) specified. If the file isn't found in the [include_path](https://www.php.net/manual/en/ini.core.php#ini.include-path), include will finally check in the calling script's own directory and the current working directory before failing.

https://www.php.net/manual/en/function.include.php
>>>

Since I don't want / I'm unable to modify all the `require_once 'Zend/[...]';` out there I decided to brute force things with a bunch of `include-path` in my `composer.json` - https://getcomposer.org/doc/04-schema.md#include-path

```json
  "include-path": [
    "vendor/zf1s/zend-acl/library",
    "vendor/zf1s/zend-application/library",
    "vendor/zf1s/zend-auth/library",
    "vendor/zf1s/zend-cache/library",
    "vendor/zf1s/zend-config/library",
    "vendor/zf1s/zend-console-getopt/library",
    "vendor/zf1s/zend-controller/library",
    "vendor/zf1s/zend-crypt/library",
    "vendor/zf1s/zend-date/library",
    "vendor/zf1s/zend-db/library",
    "vendor/zf1s/zend-dom/library",
    "vendor/zf1s/zend-exception/library",
    "vendor/zf1s/zend-file-transfer/library",
    "vendor/zf1s/zend-filter/library",
    "vendor/zf1s/zend-form/library",
    "vendor/zf1s/zend-json/library",
    "vendor/zf1s/zend-layout/library",
    "vendor/zf1s/zend-loader/library",
    "vendor/zf1s/zend-locale/library",
    "vendor/zf1s/zend-log/library",
    "vendor/zf1s/zend-measure/library",
    "vendor/zf1s/zend-navigation/library",
    "vendor/zf1s/zend-paginator/library",
    "vendor/zf1s/zend-registry/library",
    "vendor/zf1s/zend-server/library",
    "vendor/zf1s/zend-session/library",
    "vendor/zf1s/zend-test/library",
    "vendor/zf1s/zend-translate/library",
    "vendor/zf1s/zend-uri/library",
    "vendor/zf1s/zend-validate/library",
    "vendor/zf1s/zend-version/library",
    "vendor/zf1s/zend-view/library",
    "vendor/zf1s/zend-xml/library"
  ],
```

And it works 🥳 but it's ugly 💩 

In fact, the original Zend composer definition had an `include-path` statement - https://github.com/zendframework/zf1/blob/master/composer.json#L19

So, my hypothesis here is to simply provide an equivalent definition, both for the monorepo (which I don't use in my example) and every single modules.

**Comment by falkenhawk** [comment]
Yes @fredericgboutin-yapla that's exactly what you need to do if you need to combine zf1s with zf1 extensions which still utilize require_once... and yes it is ugly and potentially may affect general php performance regarding file lookups (although I am not sure about that, only guessing) so I would not want it to be added to the main composer.json of zf1s so that it affects everyone.

It's a matter of finding a compromise - either forking/patching old codebase for zf1 extensions, or adding include-path for all components in your project. Or switching to zf1-future, diablomedia, etc. forks :)

Again, thanks for sharing!
